File: src/main/java/rag/trace/JsonlTraceSink.java
package rag.trace;

import java.io.*;

// concrete implementation of TraceSink that writes events to a JSON Lines format
public class JsonlTraceSink implements TraceSink {

    private final FileWriter writer;

    public JsonlTraceSink(String path) throws IOException {
        this.writer = new FileWriter(path, true);
    }

    @Override
    public void record(TraceEvent e) throws RuntimeException {
        try {

            // manual escaping
            String summary = e.summary == null ? "" : e.summary.replace("\"", "\\\"");
            String error = e.error == null ? "" : e.error.replace("\"", "\\\"");

            // generate a JSON file
            String json = String.format(
                "{\"timestamp\":%d,\"stage\":\"%s\",\"summary\":\"%s\",\"durationMs\":%d,\"error\":\"%s\"}\n",
                e.timestamp,
                e.stage,
                summary,
                e.durationMs,
                error
            );

            writer.write(json);
            writer.flush();
        } catch (IOException ex) {
            // log the error
            System.err.println("failed regarding to I/O while writing to JsonlTraceSink" + ex.getMessage());
        }
    }
}

File: src/main/python/rag/trace/jsonl_trace_sink.py
from typing import TextIO
from pathlib import Path
from rag.trace.trace_sink import TraceSink
from rag.trace.trace_event import TraceEvent


class JsonlTraceSink(TraceSink):
    def __init__(self, path: str) -> None:
        self.writer: TextIO = open(path, 'a', encoding='utf-8')

    def accept(self, e: TraceEvent) -> None:
        try:
            summary: str = (e.summary.replace('"', '\\"') if e.summary else "")
            error: str = (e.error.replace('"', '\\"') if e.error else "")
            json_str: str = (
                f'{{"timestamp":{e.timestamp},"stage":"{e.stage}",'
                f'"summary":"{summary}","durationMs":{e.duration_ms},'
                f'"error":"{error}"}}\n'
            )
            self.writer.write(json_str)
            self.writer.flush()
        except Exception as ex:
            raise RuntimeError(str(ex)) from ex






